lang/step.md

```diff
@@ -56,13 +56,15 @@ impl<M: Memory> Machine<M> {
             Constant::Int(i) => Value::Int(i),
             Constant::Bool(b) => Value::Bool(b),
             Constant::Tuple(args) => {
-                let args = args.into_iter()
-                    .map(|arg| self.eval_constant(arg))
-                    .collect();
-                Value::Tuple(args)
+                let mut vals = Vec::new();
+                for arg in args {
+                    let val = self.eval_constant(arg)?;
+                    vals.push(val);
+                }
```

Comment:
I'd rather use a functional `collect` here than an imperative `push` loop. Can't we get that to work?

Reply:
agreed, we cannot do `.map(|c| self.eval_constant(c)?)` though, as the `?` will try to return from the closure, not our outer function.
Anything else I've come up with, just wasn't easier than this imperative solution.

Reply:
nvm, `try_collect` is better